### PR TITLE
DOC: Clarify groupby.first does not use nulls 

### LINKED ISSUE #8427: BUG: groupby.first/last with nans
This is incorrect, as this is applied column by column (as they are different dtypes)
so `_first_compat` should first compute the mask then use it.

from [SO](http://stackoverflow.com/questions/26108181/selecting-first-row-with-groupby-and-nan-columns)

```
In [18]: df = pd.DataFrame([{'id':"a",'val':np.nan, 'val2':-1},{'id':"a",'val':'TREE','val2':15}])

In [19]: df
Out[19]: 
  id   val  val2
0  a   NaN    -1
1  a  TREE    15

In [20]: df.groupby('id').first()
Out[20]: 
     val  val2
id            
a   TREE    -1
```

### PR DESCRIPTION
**Pull request by NumberPiOso**
- [X] closes  #8427
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).

### CODE REVIEW

**Comment by jreback** [comment]
great. can you also update `.last()`?

**Comment by NumberPiOso** [comment]
Nowadays, when I run the same .first() example I get the following error. Do you know the reason? I will still commit what I would think would be the change for .last()

```python
In [4]: df = pd.DataFrame(dict(A=[1, 1, 3], B=[None, 5, 6], C=[1, 2, 3]))

In [5]: df.groupby("A").first()
<ipython-input-5-b69697f6536c>:1: FutureWarning: Dropping invalid columns in DataFrameGroupBy.first is deprecated. In a future version, a TypeError will be raised. Before calling .first, select only columns which should be valid for the function.
  df.groupby("A").first()
Out[5]: 
Empty DataFrame
Columns: []
Index: [1, 3]
```

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -2225,8 +2225,37 @@ def max(
...
+            Include only float, int, boolean columns. If None, will attempt to use
+            everything, then use only numeric data.
+        min_count : int, default -1
+            The required number of valid values to perform the operation. If fewer
+            than ``min_count`` non-NA values are present the result will be NA.
+
+        Returns
+        -------
+        Series or DataFrame
+            First not non-null of values within each group.
```
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:2246]: Just "First non-null"

File: pandas/core/groupby/groupby.py [hunk 2]
```diff
@@ -2250,8 +2279,37 @@ def first(x: Series):
...
+            Include only float, int, boolean columns. If None, will attempt to use
+            everything, then use only numeric data.
+        min_count : int, default -1
+            The required number of valid values to perform the operation. If fewer
+            than ``min_count`` non-NA values are present the result will be NA.
+
+        Returns
+        -------
+        Series or DataFrame
+            Last not non-null of values within each group.
```
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:2300]: Ditto.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Small request, otherwise looks great.

File: pandas/core/groupby/groupby.py [hunk 3]
```diff
@@ -2250,8 +2285,44 @@ def first(x: Series):
...
+        --------
+        >>> df = pd.DataFrame(dict(A=[1, 1, 3], B=[5, None, 6], C=[1, 2, 3]))
+        >>> df.groupby("A").last()
+             B  C
+        A
+        1  5.0  2
+        3  6.0  3
+
+        """
+
```
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:2325]: Need to remove this blank line - causes docstring validation to fail.

**Review by rhshadrach — CHANGES_REQUESTED** [review]

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Thanks @NumberPiOso!

**Pull request merged**